vllm/v1/sample/sampler.py

```diff
@@ -166,3 +166,14 @@ def apply_penalties(
                 sampling_metadata.repetition_penalties,
                 sampling_metadata.output_token_ids)
         return logits
+
+    def apply_logits_bias(
+        self,
+        logits: torch.Tensor,
+        sampling_metadata: SamplingMetadata,
+    ) -> torch.Tensor:
+        for i, logit_bias in enumerate(sampling_metadata.logit_bias):
+            if logit_bias is not None:
+                for token_id, bias in logit_bias.items():
+                    logits[i, token_id] += bias
```

Comment:
Can we add a comment like `TODO: This is extremely slow. Optimize this.`?

Reply:
@njhill @robertgshaw2-redhat Although this implementation is a bit slow, I'm comfortable merging the PR since I haven't found a way to optimize it yet, and getting the functionality in is our top priority. What do you think?

Reply:
Yeah, I agree. We can write some customized kernel to handle such things in c++. Also we may change the representation of logit_bias from dict to key value pair.

I can create some TODO as follow up.

Reply:
@houseroad Sg. Could you please update the PR?

Reply:
@WoosukKwon I think that's fine since we need the functionality asap. But I think it should be simple to vectorize this without any custom kernel.

We just need to maintain in the batch three one-dim tensors of the same length: 
- all the bias values concatenated (b)
- corresponding request indices (s)
- corresponding token ids (t)

We only need to update these when any requests with logit bias are added or removed from the batch.

Then we can just do `logits[(s, t)] += v`

Reply:
Updated the PR. Please let me know if you want me directly jump to the optimized solution.

Reply:
@njhill @houseroad Considering that the code is pretty isolated, I think we can merge the PR first and have a followup PR to optimize it.

Reply:
@njhill Re your idea: Each request's `logits_bias` has different lengths. How do we handle that (with the persistent batch)?

Reply:
I think we can have ragged format representation, 3 tensors, tensor a for length/offset of each batch, tensor b for token ids, tensor c for bias. And one torch op takes these 3 tensors as inputs, and we leverage C++ logic to handle it. It should be much faster then the current logic.

Maybe in the SamplingMeta or param, we should just preproces things like this.

The additional overhead is that once we update the batch, we may need to generate new tensors, which should be acceptable.